test/hotspot/jtreg/compiler/onSpinWait/TestOnSpinWaitAArch64.java

```diff
@@ -0,0 +1,189 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test TestOnSpinWaitAArch64
+ * @summary Checks that java.lang.Thread.onSpinWait is intrinsified with instructions specified with '-XX:OnSpinWaitInst' and '-XX:OnSpinWaitInstCount'
+ * @bug 8186670
+ * @library /test/lib
+ *
+ * @requires vm.flagless
+ * @requires os.arch=="aarch64"
+ *
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c2 nop 7
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c2 isb 3
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c2 yield 1
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c1 nop 7
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c1 isb 3
+ * @run driver compiler.onSpinWait.TestOnSpinWaitAArch64 c1 yield
+ */
+
+package compiler.onSpinWait;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.ListIterator;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class TestOnSpinWaitAArch64 {
+    public static void main(String[] args) throws Exception {
+        String compiler = args[0];
+        String spinWaitInst = args[1];
+        String spinWaitInstCount = (args.length == 3) ? args[2] : "1";
+        ArrayList<String> command = new ArrayList<String>();
+        command.add("-XX:+IgnoreUnrecognizedVMOptions");
+        command.add("-showversion");
+        command.add("-XX:-BackgroundCompilation");
+        command.add("-XX:+UnlockDiagnosticVMOptions");
+        command.add("-XX:+PrintAssembly");
+        if (compiler.equals("c2")) {
+            command.add("-XX:-TieredCompilation");
+        } else if (compiler.equals("c1")) {
+            command.add("-XX:+TieredCompilation");
+            command.add("-XX:TieredStopAtLevel=1");
+        } else {
+            throw new RuntimeException("Unknown compiler: " + compiler);
+        }
+        command.add("-Xbatch");
+        command.add("-XX:OnSpinWaitInst=" + spinWaitInst);
+        command.add("-XX:OnSpinWaitInstCount=" + spinWaitInstCount);
+        command.add("-XX:CompileCommand=compileonly," + Launcher.class.getName() + "::" + "test");
+        command.add(Launcher.class.getName());
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(command);
+
+        OutputAnalyzer analyzer = new OutputAnalyzer(pb.start());
+
+        analyzer.shouldHaveExitValue(0);
+
+        System.out.println(analyzer.getOutput());
+
+        checkOutput(analyzer, getSpinWaitInstHex(spinWaitInst), Integer.parseInt(spinWaitInstCount));
+    }
+
+    private static String getSpinWaitInstHex(String spinWaitInst) {
+      if ("nop".equals(spinWaitInst)) {
+          return "1f20 03d5";
```

Comment:
I'm getting the following when running these tests:
STDERR:
java.lang.RuntimeException: Wrong instruction 1f20 03d5 count 0!
  -- expecting 7
        at compiler.onSpinWait.TestOnSpinWaitAArch64.checkOutput(TestOnSpinWaitAArch64.java:163)
        at compiler.onSpinWait.TestOnSpinWaitAArch64.main(TestOnSpinWaitAArch64.java:82)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
        at java.base/java.lang.Thread.run(Thread.java:833)

I have hsdis installed as a matter of course, is the test written assuming hsdis is not present?

Reply:
Yes, currently it is assumed that hsdis is not present and hence that the disassembly produces hex insns. However, if you install hsdis as part of the build (as I always do) then these tests will fail.  It would really be better if the test looked for the correct hex insns or disassembled insns.

Reply:
I'd suggest the most straightforward way to deal with this would be to check for the hex and fail if neither are found. An alternative could be to switch behaviour if the message "Loaded disassembler from hsdis-aarch64.so" is found.

Reply:
I guess so. This should be fixed.

Reply:
I have a fix which makes the test working with or without hsdis. I am testing it.

Reply:
Done.

Reply:
The [IR test framework](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md) can parse the C2 opto assembly output, and also control compilation level and flags through method annotations. I'm not sure if we could use that to check the C1 output though. Maybe @chhagedorn has some ideas?

Reply:
The framework only supports regex matching on C2's ideal and opto assembly output. You would still need the current checks for C1's output. So, I'm not sure if it's worth transforming the C2 checks only to use the framework.

Reply:
You might be over-thinking this a little. You already check that the assembler macro does the right thing, so as long as you've made sure the C1 code actually calls the macro you're good. Also, this is a performance-only change (does not affect correctness) and it may not have a great deal of affect in C1-generated code anyway.

Reply:
@nick-arm Thank you for the link.